rclcpp_components/src/component_manager_isolated.cpp

```diff
@@ -0,0 +1,153 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rclcpp_components/component_manager_isolated.hpp"
+
+#include <memory>
+#include <string>
+#include <utility>
+
+namespace rclcpp_components
+{
+
+ComponentManagerIsolated::~ComponentManagerIsolated()
+{
+  if (node_wrappers_.size()) {
+    for (auto & prom : dedicated_promises_) {
+      prom.second.set_value();
+    }
+    for (auto & exec : dedicated_executors_) {
+      exec.second->cancel();
+    }
+    for (auto & t : dedicated_threads_) {
+      t.second.join();
+    }
+    node_wrappers_.clear();
+  }
+}
+
+void
+ComponentManagerIsolated::on_load_node(
+  const std::shared_ptr<rmw_request_id_t> request_header,
+  const std::shared_ptr<LoadNode::Request> request,
+  std::shared_ptr<LoadNode::Response> response)
+{
+  (void) request_header;
```

Comment:
This entire code block L46-L87 and L103-114 are exactly the same as in the other container. 

What you could do instead is what I suggested above. Put the things in L88-L102 into a separate, virtual function with the implementation that is overrided. Then you won't even need to override or implement `on_load_node`. This is really good not to duplicate code + makes it so any change to the main container are automatically applied here. 

If you did that too, it would resolve Ivan's comment above on duplication as well, because you wouldn't have `on_load_node` at all.